[frymash] iP

frymash opened the pull request:
# Luke

> "You'll find I'm full of surprises." - Luke Skywalker


*  **Luke is a chatbot that works like a to-do list 📝** It supports the following event types:
1. todo
2. deadline
3. event

The to-do list looks like this:
- [ ] _todo_ borrow book
- [x] _deadline_ submit PE by 1/9/2024 14:00
- [ ] _event_ sports day from 2/9/2024 10:00 to 13:00

Here's some sample code from Luke:
```java
    /**
     * Marks or unmarks tasks
     * @param command the mark/unmark command word and its associated tasks
     */
    public static String handleTaskMarking(Command command) {
        int taskToMark = Integer.parseInt(command.getArgs());
        try {
            Task task = taskList.getTask(taskToMark - 1);
            task.changeMark();
            Storage.saveData(taskList);
            return (command.getCommand().equals("mark")
                    ? "ok i've marked"
                    : "ok i've unmarked")
                    + " this task:\n"
                    + Constants.INDENT + task.taskDescription();

        } catch (NumberFormatException | IndexOutOfBoundsException e) {
            return TaskList.taskNotFound(taskToMark);
        }
    }
```


You can run Luke by [cloning its repo](github.com/frymash/ip) and running Gradle.
``gradlew run``

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

frymash committed:
Rename file and add greeting

frymash committed:
Add echo functionality

frymash committed:
Implement add-to-list functionality

frymash committed:
Implement list marking functionality

frymash committed:
Implement Task subclasses (Todo, Deadline, Event)

frymash committed:
Handle incorrect inputs

frymash committed:
Add support for deleting tasks

frymash committed:
Implement semi-automated testing

frymash committed:
Refactor addToList: replace if-else w/ switch-case

frymash committed:
Remove recursive calls to acceptCommand

frymash committed:
Fix import lines' code style in Bot

frymash committed:
Fix style: unindent switch-case statements

frymash committed:
Add support for loading task lists from .txt

frymash committed:
Fix bug: bot exits after creating task list txt

frymash committed:
Implement task autosaving functionality

frymash committed:
Add new test inputs

frymash committed:
Merge branch 'branch-Level-7'
The master branch does not have the functionality to
1. save tasks in the hard disk automatically
2. load the data from the hard disk when the chatbot starts up

In branch-Level-7, the features mentioned above have been implemented.

Hence, we must merge branch-Level-7 into master so that the functionality
can also be available on master.

The merge into master must occur to recognise the significance of this
new functionality to the chatbot as a whole. Without the merge,
further development of the chatbot will only continue on branch-Level-7.

frymash committed:
Remove wildcard import for java.nio

frymash committed:
Convert Deadline dates to LocalTimeDates

frymash committed:
Merge branch 'branch-Level-8'
In the Deadline class, dates are stored as Strings. Strings may not be the best type to represent dates given that there are Java libraries (such as the time library) which can represent dates more appropriately.

Let's use java.time.LocalDateTime to represent dates internally instead.

It is a type that is specifically designed to represent a date and deal with date-related operations.

frymash committed:
Refactor code into new classes

frymash committed:
Organise files into packages

frymash committed:
Merge branch 'add-packages'
/src/java consists of Java files that are essential for the chatbot to function. However, these files have not been sorted into folders.

Therefore, it is not clear which files share a closer relationship as compared to others.

To reflect these relationships more clearly, let's organise all the files into suitable Java packages.

These packages "encapsulate" the different files and make connections between related files clearer.

frymash committed:
Merge branch 'add-gradle-support'

frymash committed:
Create basic JUnit tests

frymash committed:
Refine style for package/import statements

frymash committed:
Configure files for jar creation

frymash committed:
Merge branch 'create-jar-file'

frymash committed:
Integrate JavaDoc into Storage, Parser, Ui

frymash committed:
Comply with SE-EDU coding standards

frymash committed:
Implement searching feature

frymash committed:
Merge tag 'A-JavaDoc'

frymash committed:
Merge tag 'A-CodingStandard'

frymash committed:
Resolve merge conflicts

frymash committed:
Intro checkstyle and make adjustments

frymash committed:
Merge branch 'A-Checkstyle'

frymash committed:
Fix style issues in TaskList.java

frymash committed:
Merge branch 'branch-A-Checkstyle'

frymash committed:
Create new methods to add each task type to list

frymash committed:
Change all system printouts to string returns

frymash committed:
Add files that contribute to JavaFX integration

frymash committed:
Construct a JavaFX GUI for the chatbot

frymash committed:
Merge branch 'branch-Level-10'

frymash committed:
Clean up code residue

frymash committed:
Clean up code further

Raghava-Chittidi commented on src/main/java/luke/task/UnknownCommandException.java:
```diff
@@ -0,0 +1,7 @@
+
+public class UnknownCommandException extends Exception{
+    public UnknownCommandException(){
+        super();
```
Maybe an unknown command exception message can be set at this centralized location so that in other places it does not have to be set again?

Raghava-Chittidi commented on src/main/java/luke/task/Deadline.java:
```diff
@@ -0,0 +1,39 @@
+import java.time.format.DateTimeFormatter;
+
+public class Deadline extends Task {
+    protected static final DateTimeFormatter DATE_TIME_INPUT_FORMATTER
+            = DateTimeFormatter.ofPattern("d/MM/yyyy HH:mm");
+    protected static final DateTimeFormatter DATE_TIME_OUTPUT_FORMATTER
+            = DateTimeFormatter.ofPattern("MMM d yyyy, HH:mm");
```
Maybe this can be extracted out to a DateTime class so that the Event class can also use this?

Raghava-Chittidi commented on src/main/java/luke/env/Constants.java:
```diff
@@ -0,0 +1,11 @@
+
+import java.util.List;
+
+public class Constants {
```
I like how you have an env folder for all your constants

Raghava-Chittidi commented on src/main/java/luke/Storage.java:
```diff
@@ -0,0 +1,58 @@
+import luke.task.TaskList;
+
+public class Storage {
+    private static final Path saveDataPath = Paths.get(Constants.FILE_PATH);
```
Should this be capitalized, since it is a constant?

Raghava-Chittidi reviewed (commented):
Overall, the code LGTM! It is very well done with just a few minor nits

JoanneHing commented on src/main/java/luke/DialogBox.java:
```diff
@@ -0,0 +1,56 @@
+import javafx.scene.image.ImageView;
+import javafx.scene.layout.HBox;
+
+public class DialogBox extends HBox {
```
The overall class look good! But the JavaDoc for class is missed out, perhaps can add on to that!

JoanneHing commented on src/main/java/luke/DialogBox.java:
```diff
@@ -0,0 +1,56 @@
+        setAlignment(Pos.TOP_LEFT);
+    }
+
+    public static DialogBox getUserDialog(String text, Image img) {
+        return new DialogBox(text, img);
+    }
+
+    public static DialogBox getLukeDialog(String text, Image img) {
```
The header comments might be missed out for this two functions too?

JoanneHing commented on src/main/java/luke/Luke.java:
```diff
@@ -0,0 +1,10 @@
+package luke;
+
```
Perhaps can add JavaDoc for this class too?

JoanneHing commented on src/main/java/luke/LukeUiWindow.java:
```diff
@@ -0,0 +1,118 @@
+                    response = "save file created! ok, i'm all ears now. tell me what you need.";
+                } catch (IOException e) {
+                    response = "oof, i couldn't create the file. i'll exit first - try restarting me!";
+                    // System.exit(0);
```
Maybe the unused code can be removed instead of commented?

frymash replied:
Ooh, I missed this one out. Thank you!

JoanneHing commented on src/main/java/luke/LukeUiWindow.java:
```diff
@@ -0,0 +1,118 @@
+            }
+            case "n", "no" -> {
+                response = "alright then. cya ;)";
+                // System.exit(0);
```
same here as above :D

JoanneHing commented on src/main/java/luke/NoSaveDataFoundException.java:
```diff
@@ -0,0 +1,7 @@
+package luke;
+
```
Oops, JavaDoc is missing here also :O

JoanneHing commented on src/main/java/luke/Parser.java:
```diff
@@ -0,0 +1,34 @@
+import java.util.Arrays;
+
+import luke.command.Command;
+public class Parser {
```
Same for this class :D

JoanneHing commented on src/main/java/luke/command/Command.java:
```diff
@@ -0,0 +1,31 @@
+package luke.command;
+
```
This class is missing of JavaDoc too

JoanneHing commented on src/main/java/luke/env/Constants.java:
```diff
@@ -0,0 +1,11 @@
+package luke.env;
+
+import java.util.List;
+
```
The idea having a env class for the constants is good organization!

JoanneHing commented on src/main/java/luke/task/Event.java:
```diff
@@ -0,0 +1,25 @@
+package luke.task;
+
```
Maybe can considering using DateTime class for event's timing also

JoanneHing commented on src/main/java/luke/task/Task.java:
```diff
@@ -0,0 +1,24 @@
+                ? "[X]"
+                : "[ ]";
+    }
+
```
Good idea to create abstract function to ensure the children class implement the functions!

JoanneHing commented on src/main/java/luke/task/TaskList.java:
```diff
@@ -0,0 +1,170 @@
+    }
+
+    private String addDeadlineToList(String taskDetails, boolean isMarked, boolean isLoadingFromDisk) {
+        String[] taskAndDeadline;
+        if (isLoadingFromDisk) {
+            taskAndDeadline = taskDetails.split(" by ");
+        } else {
+            taskAndDeadline = taskDetails.split(" /by ");
+        }
+        String taskName = taskAndDeadline[0];
+        String deadline = taskAndDeadline[1];
+        Deadline dl = new Deadline(taskName, deadline, isMarked);
+        addTask(dl);
+        if (!isLoadingFromDisk) {
+            return "the new deadline's been added to your task list:\n"
+                    + Constants.INDENT + dl.taskDescription() + "\n"
+                    + listSizeUpdateMessage();
+        }
+        return "";
+    }
+
+    private String addEventToList(String taskDetails, boolean isMarked, boolean isLoadingFromDisk) {
+        String[] taskAndTimings;
+        if (isLoadingFromDisk) {
+            taskAndTimings = taskDetails.split(" from | to ");
+        } else {
+            taskAndTimings = taskDetails.split(" /from | /to ");
+        }
+        String taskName = taskAndTimings[0];
+        String from = taskAndTimings[1];
+        String to = taskAndTimings[2];
+        Event event = new Event(taskName, from, to, isMarked);
+        addTask(event);
+        if (!isLoadingFromDisk) {
+            return "aaaaand this event is now in your task list:\n"
+                    + Constants.INDENT + event.taskDescription() + "\n"
+                    + listSizeUpdateMessage();
+        }
+        return "";
+    }
```
Maybe can extract the function for parsing data to their respective task class. For example you can parse the input into two parts, the task type and the description, and process the description for deadline and event task to split the task description and timing within the class constructor. Just a small suggestion!

JoanneHing reviewed (commented):
Overall LGTM! Just that some of the class has missing JavaDoc for the class and the public methods, it would be nice if the JavaDoc could be added! Good job!

frymash committed:
Add assertions and remove dead lines

frymash committed:
Modify assertions, add pauses before exits

frymash committed:
Remove handleMissingFile, update save features

frymash committed:
Updated bot responses to malformed commands

frymash committed:
Add JavaDoc comments to classes and functions

frymash committed:
Merge branch 'branch-A-Assertions'
Most of the classes and methods in Luke do not have Javadoc comments.

Even though Luke remains functional without these comments, the absence of such comments remains problematic as the codebase may be difficult to read without them.

To address this, Javadoc comments have been added to the majority of classes and methods in Luke (save for trivial methods like setters and getters).

frymash committed:
Merge branch-A-CodeQuality

frymash committed:
Implement duplicate task detection

Luke is not able to detect duplicate tasks; two tasks of the same name can be added to the task list.

Assuming a duplicate task was added by accident, the user will not be notified and the mistake will likely go undetected.

To resolve this issue, let's implement duplicate task detection in TaskList. Any two tasks will be considered duplicates of one another if (1) their names AND types are the same OR (2) they reference the same location in memory. Additionally, Deadline tasks are now formatted to consistently store dates and times in a uniform format, regardless of the format used in the input.

Duplicate task detection should be implemented in TaskList and not other parts of the Luke codebase. This will ensure we can catch duplicate tasks as soon as Luke attempts to them to the TaskList.

frymash committed:
Merge branch 'branch-C-DetectDuplicates'

frymash committed:
Set GUI title and display bye message correctly

frymash committed:
Merge branch 'final-refinements'

frymash committed:
Patch number formatting bug in handleCommand

frymash committed:
Update README.md with info on Luke's commands

The README for Luke still contains outdated information from the Duke project template.

It does not explain what Luke is and how it can be used.

Let's update the README with information on how to use the commands available in Luke (todo, deadline, event, delete, find, list, mark, unmark, and bye) so that users can learn how to use Luke.

frymash committed:
Update handleTaskMarking Javadoc comment

frymash committed:
Upload representative screenshot of Luke

frymash committed:
Update README.md with additional details

frymash committed:
Merge branch 'final-refinements'

frymash committed:
Merge branch 'branch-A-UserGuide'

frymash committed:
Merge pull request #3 from frymash/branch-A-UserGuide

Update README.md with additional details

frymash committed:
Update README.md in /docs/ to mirror ip/README.md

frymash committed:
Update links to images in /docs/README.md

frymash committed:
Update links to images in /docs/README.md

frymash committed:
Update links to images in /docs/README.md

GitHub Pages isn't displaying the images correctly, so we have to edit the images' URLs in README.md to fix this problem.

frymash committed:
Revise Luke save file sample

frymash committed:
Update links to images in /docs/README.md
